File: source-app/backend/app/services/unit_resolution_service.py
"""Canonical wholesale unit + package-size resolution (backend SSOT).

Package tokens such as ``50KG`` / ``850GM`` / ``15LTR`` are **sizes**, not selling units.
"""

from __future__ import annotations

import json
import re
from dataclasses import asdict, dataclass
from decimal import Decimal
from functools import lru_cache
from pathlib import Path
from typing import Any

from app.models.catalog import CatalogItem
from app.services.trade_unit_type import parse_kg_per_bag_from_name

_SIZE_KG = re.compile(r"(\d+)\s*KG", re.IGNORECASE)
_SIZE_GM = re.compile(r"(\d+)\s*GM", re.IGNORECASE)
_SIZE_LTR = re.compile(r"(\d+)\s*LTR", re.IGNORECASE)
_SIZE_ML = re.compile(r"(\d+)\s*ML", re.IGNORECASE)


@dataclass(frozen=True)
class UnitResolution:
    selling_unit: str
    stock_unit: str | None = None
    display_unit: str | None = None
    package_type: str | None = None
    package_size: Decimal | None = None
    package_measurement: str | None = None
    conversion_factor: Decimal = Decimal("1")
    confidence: Decimal = Decimal("0")
    rule_id: str | None = None
    canonical_unit_type: str | None = None

    def as_dict(self) -> dict[str, Any]:
        d = asdict(self)
        for k in ("package_size", "conversion_factor", "confidence"):
            v = d.get(k)
            if isinstance(v, Decimal):
                d[k] = float(v) if v is not None else None
        canon = (self.canonical_unit_type or self.selling_unit or "").upper()
        d["canonical_unit_type"] = canon or None
        d["inferred_confidence"] = float(self.confidence) if self.confidence is not None else None
        d["unit_profile_source"] = self.rule_id
        if (
            (self.selling_unit or "").upper() == "BAG"
            and self.package_measurement == "KG"
            and self.package_size is not None
        ):
            d["kg_per_bag"] = float(self.package_size)
        else:
            d["kg_per_bag"] = None
        return d


@lru_cache(maxsize=1)
def _rules() -> dict[str, Any]:
    path = Path(__file__).with_name("unit_rules_master.json")
    return json.loads(path.read_text(encoding="utf-8"))


def _parse_size_tokens(upper_name: str) -> tuple[Decimal | None, str | None]:
    if m := _SIZE_KG.search(upper_name):
        return Decimal(m.group(1)), "KG"
    if m := _SIZE_GM.search(upper_name):
        return Decimal(m.group(1)), "GM"
    if m := _SIZE_LTR.search(upper_name):
        return Decimal(m.group(1)), "LTR"
    if m := _SIZE_ML.search(upper_name):
        return Decimal(m.group(1)), "ML"
    return None, None


def _apply_result(
    upper_name: str,
    result: dict[str, Any],
    rule_id: str,
) -> UnitResolution:
    su = str(result.get("selling_unit") or "").upper()
    size, meas = _parse_size_tokens(upper_name)
    raw_pt = result.get("package_type")
    pt = str(raw_pt).upper() if raw_pt else None
    if isinstance(pt, str) and not pt.strip():
        pt = None
    cf_raw = result.get("conversion_factor")
    cf_explicit = Decimal(str(cf_raw)) if cf_raw is not None else None
    st = result.get("stock_unit")
    stock = str(st).upper() if st else None
    parsed_pt, stock2, cf2 = _infer_stock_and_conversion(su, size, meas, pt)
    conv_final = cf_explicit if cf_explicit is not None else cf2
    canon = su if su else None
    return UnitResolution(
        selling_unit=su,
        stock_unit=stock or stock2,
        display_unit=None,
        package_type=parsed_pt,
        package_size=size,
        package_measurement=meas,
        conversion_factor=conv_final,
        confidence=Decimal("85"),
        rule_id=rule_id,
        canonical_unit_type=canon,
    )


def _infer_stock_and_conversion(
    selling_unit: str,
    size: Decimal | None,
    meas: str | None,
    package_type: str | None,
) -> tuple[str | None, str | None, Decimal]:
    stock = "PCS"
    conv = Decimal("1")
    pt = package_type
    if selling_unit == "BAG" and size is not None and meas == "KG":
        stock = "KG"
        conv = size
        pt = pt or "SACK"
    elif selling_unit == "TIN" and size is not None and meas in ("LTR", "ML"):
        stock = "TIN"
        conv = Decimal("1")
        pt = pt or "TIN"
    elif selling_unit == "BOX" and size is not None:
        stock = "PCS"
        conv = Decimal("1")
        pt = pt or "BOX"
    elif selling_unit == "KG":
        stock = "KG"
        conv = Decimal("1")
        pt = pt or "LOOSE"
    return pt, stock, conv


def _category_rule_matches_key(cat_upper: str, key: str) -> bool:
    ku = key.upper().replace("_", " ").replace("-", " ")
    if ku in cat_upper or cat_upper in ku:
        return True
    ku_compact = ku.replace(" ", "")
    cc = cat_upper.replace(" ", "")
    if ku_compact and ku_compact in cc:
        return True
    return False


def _match_condition(
    upper_name: str,
    upper_cat: str,
    brand_detected: bool,
    cond: dict[str, Any],
) -> bool:
    any_tokens = [str(x).upper() for x in (cond.get("contains_any") or [])]
    compact_name = upper_name.replace(" ", "")
    if any_tokens and not any(t in upper_name or t.replace(" ", "") in compact_name for t in any_tokens):
        return False
    excludes = [str(x).upper() for x in (cond.get("excludes_any") or [])]
    if excludes and any(t in upper_name for t in excludes):
        return False
    cats = [str(x).upper() for x in (cond.get("category_any") or [])]
    if cats and not any(c in upper_cat or upper_cat == c or _category_rule_matches_key(upper_cat, c) for c in cats):
        return False
    if cond.get("brand_detected") is True and not brand_detected:
        return False
    rx = cond.get("name_regex")
    if rx:
        try:
            if not re.search(str(rx), upper_name, re.IGNORECASE):
                return False
        except re.error:
            return False
    return True


def resolve_from_text(
    item_name: str,
    *,
    category_name: str | None = None,
    brand_detected: bool = False,
) -> UnitResolution:
    """Resolve units from free text + optional category hint (matches Flutter classifier)."""
    rules = _rules()
    upper = item_name.upper().strip()
    cat = (category_name or "").upper().strip()

    if "LOOSE" in upper:
        return UnitResolution(
            selling_unit="KG",
            package_type="LOOSE",
            stock_unit="KG",
            conversion_factor=Decimal("1"),
            confidence=Decimal("92"),
            rule_id="loose",
            canonical_unit_type="KG",
        )

    detection = rules.get("smart_detection_rules") or []
    for i, row in enumerate(detection):
        cond = row.get("condition") or {}
        result = row.get("result") or {}
        if not _match_condition(upper, cat, brand_detected, cond):
            continue
        if not result.get("selling_unit"):
            continue
        return _apply_result(upper, result, f"smart_rule_{i}")

    cat_rules = rules.get("category_rules") or {}
    for key, meta in cat_rules.items():
        if not _category_rule_matches_key(cat, key):
            continue
        du = str((meta or {}).get("default_unit") or "").upper()
        if not du:
            continue
        pt = str((meta or {}).get("package_type") or "").upper() or None
        size, meas = _parse_size_tokens(upper)
        _, stock, conv = _infer_stock_and_conversion(du, size, meas, pt)
        return UnitResolution(
            selling_unit=du,
            stock_unit=stock,
            package_type=pt,
            package_size=size,
            package_measurement=meas,
            conversion_factor=conv,
            confidence=Decimal("70"),
            rule_id=f"category_{key}",
            canonical_unit_type=du,
        )

    size, meas = _parse_size_tokens(upper)
    if size is not None and meas == "KG" and ("RICE" in upper or "SUGAR" in upper):
        pt, stock, conv = _infer_stock_and_conversion("BAG", size, meas, "SACK")
        return UnitResolution(
            selling_unit="BAG",
            stock_unit=stock,
            package_type=pt,
            package_size=size,
            package_measurement=meas,
            conversion_factor=conv,
            confidence=Decimal("65"),
            rule_id="fallback_bag_sack",
            canonical_unit_type="BAG",
        )

    return UnitResolution(
        selling_unit="PCS",
        confidence=Decimal("40"),
        rule_id="fallback_pcs",
        canonical_unit_type="PCS",
    )


def merge_unit_resolution_into_catalog_row(item: CatalogItem, ur: UnitResolution) -> None:
    """Persist smart-unit snapshot on a catalog row (caller commits)."""
    item.normalized_name = (item.name or "").strip().upper()[:512] or None
    item.selling_unit = ur.selling_unit
    item.stock_unit = ur.stock_unit
    item.display_unit = ur.display_unit or ur.selling_unit
    item.package_type = ur.package_type
    item.package_size = ur.package_size
    item.package_measurement = ur.package_measurement
    item.conversion_factor = ur.conversion_factor
    item.unit_confidence = ur.confidence
    item.smart_classification = ur.rule_id
    if ur.package_measurement == "KG" and ur.selling_unit == "BAG" and ur.package_size is not None:
        item.default_kg_per_bag = item.default_kg_per_bag or ur.package_size


def resolve_for_catalog_item(
    item: CatalogItem | None,
    *,
    item_name: str,
    category_name: str | None = None,
    brand_detected: bool = False,
) -> UnitResolution:
    """Prefer persisted ``catalog_items`` smart fields; fall back to ``resolve_from_text``."""
    name = (item.name if item is not None else None) or item_name
    cat = category_name
    text_res = resolve_from_text(name, category_name=cat, brand_detected=brand_detected)
    if item is not None and (item.selling_unit or "").strip():
        su = item.selling_unit.strip().upper()
        validation_status = (getattr(item, "validation_status", None) or "").strip().lower()
        smart_source = (getattr(item, "smart_classification", None) or "").strip().lower()
        row_is_verified = validation_status == "unit_profile_verified" or smart_source.startswith("master_item_profiles")
        # Old production rows may have persisted PCS/PIECE from raw imports. If a
        # high-confidence rule resolves the package differently and the row has
        # not been verified by the canonical profile migration, prefer the rule.
        if (
            not row_is_verified
            and su in {"PCS", "PIECE"}
            and text_res.selling_unit != su
            and text_res.confidence >= Decimal("80")
        ):
            return UnitResolution(
                selling_unit=text_res.selling_unit,
                stock_unit=text_res.stock_unit,
                display_unit=text_res.display_unit,
                package_type=text_res.package_type,
                package_size=text_res.package_size,
                package_measurement=text_res.package_measurement,
                conversion_factor=text_res.conversion_factor,
                confidence=text_res.confidence,
                rule_id=(text_res.rule_id or "text_rule") + "+override_unverified_row",
                canonical_unit_type=text_res.canonical_unit_type or text_res.selling_unit,
            )
        size = item.package_size
        meas = (item.package_measurement or "").upper() or None
        pt = (item.package_type or "").upper() if item.package_type else None
        cf = item.conversion_factor or Decimal("1")
        st = (item.stock_unit or "").upper() if item.stock_unit else None
        du = (item.display_unit or "").upper() if item.display_unit else None
        uc = item.unit_confidence or Decimal("95")
        return UnitResolution(
            selling_unit=su,
            stock_unit=st,
            display_unit=du,
            package_type=pt,
            package_size=size,
            package_measurement=meas,
            conversion_factor=cf,
            confidence=uc,
            rule_id="catalog_item_row",
            canonical_unit_type=su,
        )

    if item is not None and item.default_kg_per_bag and text_res.selling_unit == "BAG":
        kpb = item.default_kg_per_bag
        return UnitResolution(
            selling_unit=text_res.selling_unit,
            stock_unit=text_res.stock_unit or "KG",
            display_unit=text_res.display_unit,
            package_type=text_res.package_type or "SACK",
            package_size=text_res.package_size or kpb,
            package_measurement=text_res.package_measurement or "KG",
            conversion_factor=text_res.conversion_factor if text_res.package_size else kpb,
            confidence=min(text_res.confidence + Decimal("5"), Decimal("99")),
            rule_id=(text_res.rule_id or "") + "+kpb",
            canonical_unit_type=text_res.canonical_unit_type or text_res.selling_unit,
        )

    if item is not None and (item.default_kg_per_bag or parse_kg_per_bag_from_name(name)):
        kpb = item.default_kg_per_bag or parse_kg_per_bag_from_name(name)
        if kpb:
            return UnitResolution(
                selling_unit="BAG",
                stock_unit="KG",
                package_type="SACK",
                package_size=kpb,
                package_measurement="KG",
                conversion_factor=kpb,
                confidence=Decimal("72"),
                rule_id="default_kg_per_bag",
                canonical_unit_type="BAG",
            )

    return UnitResolution(
        selling_unit=text_res.selling_unit,
        stock_unit=text_res.stock_unit,
        display_unit=text_res.display_unit,
        package_type=text_res.package_type,
        package_size=text_res.package_size,
        package_measurement=text_res.package_measurement,
        conversion_factor=text_res.conversion_factor,
        confidence=text_res.confidence,
        rule_id=text_res.rule_id,
        canonical_unit_type=text_res.canonical_unit_type or text_res.selling_unit,
    )

File: new-app/backend/src/services/unitResolution.service.ts
/**
 * Canonical wholesale unit + package-size resolution.
 * Formula source: source-app/backend/app/services/unit_resolution_service.py
 */
import rulesJson from "./unit_rules_master.json";

export type UnitResolution = {
  selling_unit: string;
  stock_unit: string | null;
  display_unit: string | null;
  package_type: string | null;
  package_size: number | null;
  package_measurement: string | null;
  conversion_factor: number;
  confidence: number;
  rule_id: string | null;
  canonical_unit_type: string | null;
};

export type CatalogItemUnitFields = {
  name?: string | null;
  selling_unit?: string | null;
  stock_unit?: string | null;
  display_unit?: string | null;
  package_type?: string | null;
  package_size?: number | null;
  package_measurement?: string | null;
  conversion_factor?: number | null;
  unit_confidence?: number | null;
  validation_status?: string | null;
  smart_classification?: string | null;
  default_kg_per_bag?: number | null;
};

type RulesRoot = {
  smart_detection_rules?: Array<{
    condition?: Record<string, unknown>;
    result?: Record<string, unknown>;
  }>;
  category_rules?: Record<
    string,
    { default_unit?: string; package_type?: string }
  >;
};

const SIZE_KG = /(\d+)\s*KG/i;
const SIZE_GM = /(\d+)\s*GM/i;
const SIZE_LTR = /(\d+)\s*LTR/i;
const SIZE_ML = /(\d+)\s*ML/i;
const NN_KG = /(\d+(?:\.\d+)?)\s*KG/i;

let cachedRules: RulesRoot | null = null;

function loadRules(): RulesRoot {
  if (cachedRules) return cachedRules;
  cachedRules = rulesJson as RulesRoot;
  return cachedRules;
}

/** Formula source: trade_unit_type.py:parse_kg_per_bag_from_name */
export function parseKgPerBagFromName(itemName: string | null | undefined): number | null {
  if (!itemName) return null;
  const m = NN_KG.exec(itemName);
  if (!m) return null;
  const v = Number(m[1]);
  if (!Number.isFinite(v) || v <= 0 || v > 200) return null;
  return v;
}

export function unitResolutionAsDict(ur: UnitResolution): Record<string, unknown> {
  const canon = (ur.canonical_unit_type || ur.selling_unit || "").toUpperCase();
  const d: Record<string, unknown> = {
    selling_unit: ur.selling_unit,
    stock_unit: ur.stock_unit,
    display_unit: ur.display_unit,
    package_type: ur.package_type,
    package_size: ur.package_size,
    package_measurement: ur.package_measurement,
    conversion_factor: ur.conversion_factor,
    confidence: ur.confidence,
    rule_id: ur.rule_id,
    canonical_unit_type: canon || null,
    inferred_confidence: ur.confidence,
    unit_profile_source: ur.rule_id,
    kg_per_bag: null as number | null,
  };
  if (
    (ur.selling_unit || "").toUpperCase() === "BAG" &&
    ur.package_measurement === "KG" &&
    ur.package_size != null
  ) {
    d.kg_per_bag = ur.package_size;
  }
  return d;
}

function parseSizeTokens(upperName: string): {
  size: number | null;
  meas: string | null;
} {
  let m = SIZE_KG.exec(upperName);
  if (m) return { size: Number(m[1]), meas: "KG" };
  m = SIZE_GM.exec(upperName);
  if (m) return { size: Number(m[1]), meas: "GM" };
  m = SIZE_LTR.exec(upperName);
  if (m) return { size: Number(m[1]), meas: "LTR" };
  m = SIZE_ML.exec(upperName);
  if (m) return { size: Number(m[1]), meas: "ML" };
  return { size: null, meas: null };
}

function inferStockAndConversion(
  sellingUnit: string,
  size: number | null,
  meas: string | null,
  packageType: string | null,
): { pt: string | null; stock: string | null; conv: number } {
  let stock = "PCS";
  let conv = 1;
  let pt = packageType;
  if (sellingUnit === "BAG" && size != null && meas === "KG") {
    stock = "KG";
    conv = size;
    pt = pt || "SACK";
  } else if (sellingUnit === "TIN" && size != null && (meas === "LTR" || meas === "ML")) {
    stock = "TIN";
    conv = 1;
    pt = pt || "TIN";
  } else if (sellingUnit === "BOX" && size != null) {
    stock = "PCS";
    conv = 1;
    pt = pt || "BOX";
  } else if (sellingUnit === "KG") {
    stock = "KG";
    conv = 1;
    pt = pt || "LOOSE";
  }
  return { pt, stock, conv };
}

function categoryRuleMatchesKey(catUpper: string, key: string): boolean {
  // Formula source: unit_resolution_service.py:_category_rule_matches_key
  // Note: empty cat_upper matches every key via `cat_upper in ku` (Python/JS).
  const ku = key.toUpperCase().replace(/_/g, " ").replace(/-/g, " ");
  if (ku.includes(catUpper) || catUpper.includes(ku)) return true;
  const kuCompact = ku.replace(/ /g, "");
  const cc = catUpper.replace(/ /g, "");
  if (kuCompact && cc.includes(kuCompact)) return true;
  return false;
}

function matchCondition(
  upperName: string,
  upperCat: string,
  brandDetected: boolean,
  cond: Record<string, unknown>,
): boolean {
  const anyTokens = ((cond.contains_any as unknown[]) || []).map((x) =>
    String(x).toUpperCase(),
  );
  const compactName = upperName.replace(/ /g, "");
  if (
    anyTokens.length > 0 &&
    !anyTokens.some(
      (t) => upperName.includes(t) || compactName.includes(t.replace(/ /g, "")),
    )
  ) {
    return false;
  }
  const excludes = ((cond.excludes_any as unknown[]) || []).map((x) =>
    String(x).toUpperCase(),
  );
  if (excludes.some((t) => upperName.includes(t))) return false;
  const cats = ((cond.category_any as unknown[]) || []).map((x) =>
    String(x).toUpperCase(),
  );
  if (
    cats.length > 0 &&
    !cats.some(
      (c) =>
        upperCat.includes(c) ||
        upperCat === c ||
        categoryRuleMatchesKey(upperCat, c),
    )
  ) {
    return false;
  }
  if (cond.brand_detected === true && !brandDetected) return false;
  const rx = cond.name_regex;
  if (rx) {
    try {
      if (!new RegExp(String(rx), "i").test(upperName)) return false;
    } catch {
      return false;
    }
  }
  return true;
}

function applyResult(
  upperName: string,
  result: Record<string, unknown>,
  ruleId: string,
): UnitResolution {
  const su = String(result.selling_unit || "").toUpperCase();
  const { size, meas } = parseSizeTokens(upperName);
  const rawPt = result.package_type;
  let pt = rawPt != null ? String(rawPt).toUpperCase() : null;
  if (pt != null && !pt.trim()) pt = null;
  const cfRaw = result.conversion_factor;
  const cfExplicit = cfRaw != null ? Number(cfRaw) : null;
  const st = result.stock_unit;
  const stock = st != null ? String(st).toUpperCase() : null;
  const inferred = inferStockAndConversion(su, size, meas, pt);
  return {
    selling_unit: su,
    stock_unit: stock || inferred.stock,
    display_unit: null,
    package_type: inferred.pt,
    package_size: size,
    package_measurement: meas,
    conversion_factor: cfExplicit ?? inferred.conv,
    confidence: 85,
    rule_id: ruleId,
    canonical_unit_type: su || null,
  };
}

/** Formula source: unit_resolution_service.py:resolve_from_text */
export function resolveFromText(
  itemName: string,
  opts: { categoryName?: string | null; brandDetected?: boolean } = {},
): UnitResolution {
  const rules = loadRules();
  const upper = itemName.toUpperCase().trim();
  const cat = (opts.categoryName || "").toUpperCase().trim();
  const brandDetected = Boolean(opts.brandDetected);

  if (upper.includes("LOOSE")) {
    return {
      selling_unit: "KG",
      stock_unit: "KG",
      display_unit: null,
      package_type: "LOOSE",
      package_size: null,
      package_measurement: null,
      conversion_factor: 1,
      confidence: 92,
      rule_id: "loose",
      canonical_unit_type: "KG",
    };
  }

  const detection = rules.smart_detection_rules || [];
  for (let i = 0; i < detection.length; i++) {
    const row = detection[i]!;
    const cond = (row.condition || {}) as Record<string, unknown>;
    const result = (row.result || {}) as Record<string, unknown>;
    if (!matchCondition(upper, cat, brandDetected, cond)) continue;
    if (!result.selling_unit) continue;
    return applyResult(upper, result, `smart_rule_${i}`);
  }

  const catRules = rules.category_rules || {};
  for (const [key, meta] of Object.entries(catRules)) {
    if (!categoryRuleMatchesKey(cat, key)) continue;
    const du = String(meta?.default_unit || "").toUpperCase();
    if (!du) continue;
    const pt = String(meta?.package_type || "").toUpperCase() || null;
    const { size, meas } = parseSizeTokens(upper);
    const inferred = inferStockAndConversion(du, size, meas, pt);
    return {
      selling_unit: du,
      stock_unit: inferred.stock,
      display_unit: null,
      package_type: pt,
      package_size: size,
      package_measurement: meas,
      conversion_factor: inferred.conv,
      confidence: 70,
      rule_id: `category_${key}`,
      canonical_unit_type: du,
    };
  }

  const { size, meas } = parseSizeTokens(upper);
  if (size != null && meas === "KG" && (upper.includes("RICE") || upper.includes("SUGAR"))) {
    const inferred = inferStockAndConversion("BAG", size, meas, "SACK");
    return {
      selling_unit: "BAG",
      stock_unit: inferred.stock,
      display_unit: null,
      package_type: inferred.pt,
      package_size: size,
      package_measurement: meas,
      conversion_factor: inferred.conv,
      confidence: 65,
      rule_id: "fallback_bag_sack",
      canonical_unit_type: "BAG",
    };
  }

  return {
    selling_unit: "PCS",
    stock_unit: null,
    display_unit: null,
    package_type: null,
    package_size: null,
    package_measurement: null,
    conversion_factor: 1,
    confidence: 40,
    rule_id: "fallback_pcs",
    canonical_unit_type: "PCS",
  };
}

/** Formula source: unit_resolution_service.py:resolve_for_catalog_item */
export function resolveForCatalogItem(
  item: CatalogItemUnitFields | null,
  opts: {
    itemName: string;
    categoryName?: string | null;
    brandDetected?: boolean;
  },
): UnitResolution {
  const name = (item?.name || opts.itemName) ?? opts.itemName;
  const textRes = resolveFromText(name, {
    categoryName: opts.categoryName,
    brandDetected: opts.brandDetected,
  });

  if (item != null && (item.selling_unit || "").trim()) {
    const su = item.selling_unit!.trim().toUpperCase();
    const validationStatus = (item.validation_status || "").trim().toLowerCase();
    const smartSource = (item.smart_classification || "").trim().toLowerCase();
    const rowIsVerified =
      validationStatus === "unit_profile_verified" ||
      smartSource.startsWith("master_item_profiles");
    if (
      !rowIsVerified &&
      (su === "PCS" || su === "PIECE") &&
      textRes.selling_unit !== su &&
      textRes.confidence >= 80
    ) {
      return {
        ...textRes,
        rule_id: `${textRes.rule_id || "text_rule"}+override_unverified_row`,
        canonical_unit_type: textRes.canonical_unit_type || textRes.selling_unit,
      };
    }
    const size = item.package_size ?? null;
    const meas = (item.package_measurement || "").toUpperCase() || null;
    const pt = item.package_type ? item.package_type.toUpperCase() : null;
    const cf = item.conversion_factor ?? 1;
    const st = item.stock_unit ? item.stock_unit.toUpperCase() : null;
    const du = item.display_unit ? item.display_unit.toUpperCase() : null;
    const uc = item.unit_confidence ?? 95;
    return {
      selling_unit: su,
      stock_unit: st,
      display_unit: du,
      package_type: pt,
      package_size: size,
      package_measurement: meas,
      conversion_factor: cf,
      confidence: uc,
      rule_id: "catalog_item_row",
      canonical_unit_type: su,
    };
  }

  if (item != null && item.default_kg_per_bag && textRes.selling_unit === "BAG") {
    const kpb = item.default_kg_per_bag;
    return {
      selling_unit: textRes.selling_unit,
      stock_unit: textRes.stock_unit || "KG",
      display_unit: textRes.display_unit,
      package_type: textRes.package_type || "SACK",
      package_size: textRes.package_size ?? kpb,
      package_measurement: textRes.package_measurement || "KG",
      conversion_factor: textRes.package_size != null ? textRes.conversion_factor : kpb,
      confidence: Math.min(textRes.confidence + 5, 99),
      rule_id: `${textRes.rule_id || ""}+kpb`,
      canonical_unit_type: textRes.canonical_unit_type || textRes.selling_unit,
    };
  }

  if (item != null && (item.default_kg_per_bag || parseKgPerBagFromName(name))) {
    const kpb = item.default_kg_per_bag || parseKgPerBagFromName(name);
    if (kpb) {
      return {
        selling_unit: "BAG",
        stock_unit: "KG",
        display_unit: null,
        package_type: "SACK",
        package_size: kpb,
        package_measurement: "KG",
        conversion_factor: kpb,
        confidence: 72,
        rule_id: "default_kg_per_bag",
        canonical_unit_type: "BAG",
      };
    }
  }

  return {
    ...textRes,
    canonical_unit_type: textRes.canonical_unit_type || textRes.selling_unit,
  };
}
